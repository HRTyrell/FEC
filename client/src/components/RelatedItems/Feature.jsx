import React, {useState, useEffect} from 'react';
import checkMark from '../../assets/checkMark.png';


const Feature = ({currentProductFeatures, comparisonProductFeatures, feature}) => {

  const existsInCurrent = currentProductFeatures.includes(feature);
  const existsInComparison = comparisonProductFeatures.includes(feature);

  return (

    <div style={{
      display: 'flex',
      flexDirection: 'row',

    }}>

      {existsInCurrent ? <img src={checkMark} style={{
        display: 'flex',
        alignSelf: 'left',
        height: "1em",
        width: "1em"
      }}/> : null}

      <p style={{
        display: 'flex',
        alignSelf: 'center'
      }}>{feature.value + ' ' + feature.feature}</p>

      {existsInComparison ? <img src={checkMark}style={{
        display: 'flex',
        alignSelf: 'right',
        height: "1em",
        width: "1em"
      }}/> : null}
    </div>

  )
}
export default Feature;