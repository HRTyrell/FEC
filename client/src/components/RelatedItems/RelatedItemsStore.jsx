import create from 'zustand';


export const useRelatedItemsStore = create(set => ({
  outfitList: [],
  relatedItemsList: [],
  setOutfit: (products) => {set({outfitList: products})},
  setRelatedItems: (relatedItems) => set({relatedItemsList: relatedItems})
}))
