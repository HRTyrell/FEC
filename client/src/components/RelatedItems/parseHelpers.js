import axios from 'axios';
import {TOKEN} from '/MyConfig.js';

const options = {
  baseURL: 'http://app-hrsei-api.herokuapp.com/api/fec2/hr-rfc/',
  headers: {Authorization: TOKEN}
}

export const getRelatedProducts = productId => {
  return axios.get(`/products/${productId}/related`, options)
  .then(data => {
    let newData = data.data.map(product => {
      return getProduct(product);
    })
    return Promise.all(newData);
  })
}

export const getProduct = product => {
  return axios.get(`/products/${product}`, options).then(product => {
    return axios.get(`/products/${product.data.id}/styles`, options).then(styles => {
      product.styles = styles;
      return product;
    })}).then(product => {
      return axios.get(`/reviews/meta/?product_id=${product.data.id}`, options).then(reviews => {
        product.reviews = reviews;
        return product;
      })
    })
}


