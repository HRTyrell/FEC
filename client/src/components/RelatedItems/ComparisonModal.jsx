import {useState, useEffect} from 'react';
import styled from 'styled-components';
import Modal from 'react-modal';
import Feature from './Feature.jsx';

const customStyles = {
  overlay: {
    pointerEvents: 'none',
    background: 'none',
    zIndex: 1000,
  },
  content: {
    top: '50%',
    left: '50%',
    right: 'auto',
    bottom: 'auto',
    marginRight: '-50%',
    transform: 'translate(-50%, -50%)',

    background: 'white',
    width: '35em'
  },
};

Modal.setAppElement('#root');

const ComparisonModal = ({modalIsOpen, comparisonProduct, currentProduct }) => {
  let subtitle;

  function afterOpenModal() {
    // references are now sync'd and can be accessed.
    subtitle.style.color = '#f00';
  }
const features = comparisonProduct.features.concat(currentProduct.features);

  return (
    <div>
      <Modal
        isOpen={modalIsOpen}
        onAfterOpen={afterOpenModal}
        style={customStyles}
        contentLabel="Example Modal"
      >
        <h5 ref={(_subtitle) => (subtitle = _subtitle)}>Comparing </h5>
        <div style={{
        display: "flex",
        flexDirection: "row",
        justifyContent: "space-between",
      }}>
          <h3>{currentProduct.name}</h3>
          <h3>{comparisonProduct.name}</h3>
        </div>
        <div>
          {features.map((feature, index) => {
            return <Feature feature={feature} currentProductFeatures={currentProduct.features} comparisonProductFeatures={comparisonProduct.features} key={index}/>
          })}
        </div>
      </Modal>
    </div>
    );
}

export default ComparisonModal;