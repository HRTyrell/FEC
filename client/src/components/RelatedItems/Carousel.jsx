import React, {useState, useEffect, useRef} from 'react';
import styled from 'styled-components';
import {useRelatedItemsStore} from './RelatedItemsStore.jsx';

const CarouselStyled = styled.div`
  display: flex;
  height: 100%;
  position: relative;
  padding: 1em;
  margin: 1em;
  overflow: hidden;
`;

const CarouselControl = styled.button`
  position: sticky;
  display: flex;
  align-self: center;
  z-index: 9;
  background-color: rgba(0, 0, 0, 0.2);
  border-radius: 20%;
  border: none;
  font-size: 1.5rem;
  top: 50%;
  color: rgba(255, 255, 255, 0.5);
  cursor: pointer;

  &:hover {
    color: rgba(255, 255, 255, 1);
    background-color: rgba(0, 0, 0, 0.5);
  }`;


const CarouselControlPrev = styled(CarouselControl)`
  left: 0.5rem;

`;

const CarouselControlNext = styled(CarouselControl)`
  left: 95%;
`;
const FadedOverlay = styled.div`
position: absolute;
background: linear-gradient(to right, rgba(255,255,255,0), rgba(255,255,255,1) 99%);
border: 0px;
z-index: 3;
right: 2;
height: 100%;
// width: 100%;
max-width: 500px;
pointer-events: none;
`;

const MDiv = styled.div`
  height: 100%;
  max-width: 52em;
`;

const Carousel = (props) => {

  const { relatedItemsList, outfitList} = useRelatedItemsStore();

  const ref = useRef(null);
  const [overlay, setOverlay] = useState(true);
  const [showPrevious, setShowPrevious] = useState(false)
  const [showNext, setShowNext] = useState(false);

  useEffect(() => {
    if(ref.current.clientWidth > ref.current.parentElement.clientWidth) {
      setShowNext(true);
    }
  }, [outfitList, relatedItemsList])

  const slide = (e, shift) => {
    e.target.parentNode.scrollLeft -= shift;
    let distanceToEndOfScroll = e.target.parentNode.scrollHeight - e.target.parentNode.scrollLeft
    if (distanceToEndOfScroll <= 100) {
      setOverlay(false);
    } else {
      setOverlay(true);
      setShowNext(true);
    }
    if (e.target.parentNode.scrollLeft > 0) {
      setShowPrevious(true);
    } else {
      setShowPrevious(false);
    }
    if (distanceToEndOfScroll <= 20) {
      setShowNext(false);
    } else {
      setShowNext(true);
    }
  }

  return (
    <MDiv ref={ref}>
    {overlay ? <FadedOverlay /> : null }
    <CarouselStyled aria-label="Product Carousel">
<<<<<<< HEAD
      <CarouselControlPrev data-testid="prevButton" onClick={(e)=> slide(e, 120)}>&#8678;</CarouselControlPrev>
      <CarouselControlNext onClick={(e)=> slide(e, -120)}>&#8680;</CarouselControlNext>
=======
      {showPrevious ? <CarouselControlPrev data-testid="prevControl" onClick={(e)=> slide(e, 120)}>&#60;</CarouselControlPrev> : null }
      {showNext ? <CarouselControlNext data-testid="nextControl" onClick={(e)=> slide(e, -120)}>&#62;</CarouselControlNext> : null }
>>>>>>> main
      {React.cloneElement(props.children, {data: props.data}, null)}
    </CarouselStyled>
    </MDiv>
  )
}

export default Carousel;