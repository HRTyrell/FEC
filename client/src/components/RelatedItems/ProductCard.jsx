import {useState, useEffect} from 'react';
import styled from 'styled-components';
import { getProductStyles } from './parseHelpers.js';
import SmallStarBar from './SmallStarbar.jsx';
import ComparisonModal from './ComparisonModal.jsx';
import yellowStar from '../../assets/yellow-star.png';
import ProductStore from '../Provider/Zus_Provider.jsx';

////////////////Styles//////////////////////////////////////
const CardStyled = styled.div`
display: flex;
flex-direction: column;
border-style: solid;
border-spacing: 20px;
margin: 2px;
padding: 0;
width: 8em;
height: 18em;
`;


const ProductImageStyled = styled.div`
display: flex;
flex-direction: Column;
align-items: center;
justify-content: center;
flex-grow: 3;
margin: auto;
position: relative;
width: auto;

`;


const ImageStyled = styled.img`
display: flex;
align-items: center
position: relative;
width: 90%;
height: auto;
padding: 2%;
`;

const ButtonStyled = styled.img`
position: absolute;
top 10px;
right: 10px;
flex: none;
width: 1.5em;
`;

const OuterDescriptionDiv = styled.div`
display: flex;
flex-direction: column;
justify-content: center;
align-items: center;
width: 100%
height: 5em;
bottom: 1;
padding: 0.5em 0 0.5em 0;
background-color: rgba(211, 211, 211, 0.5);
`;
const Category = styled.p`
  margin-top: 0;
  margin-bottom: 0.7em;
  font-size: 0.7em;
`;

const ProductTitle = styled.h5`
  text-align: center;
  margin: 0 0 0 0;
  width: 100%;
`;

const Price = styled.p`
margin: 0.1em 0, 0.1em 0;
font-size: 0.9em;
`;



///////////////React Component///////////////////////////
const ProductCard = ({product}) => {

  const [modalIsOpen, setIsOpen] = useState(false);

  const {curProduct} = ProductStore();
  console.log(curProduct);

  const handleOnMouseEnter = () => {
    setTimeout(setIsOpen, 200, true);
  }

  const handleOnMouseLeave = () => {
    setTimeout(setIsOpen, 200, false);
  }


  const handleCardClick = (e) => {
    console.log('Clicked on ', product);
    console.log(product.data);
  }

  const defaultImage = product.styles.data.results[0].photos[0].thumbnail_url;

  const averageRating = () => {
    const reviewObject = product.reviews.data.ratings;
    let sumTotal = 0;
    let sumReviewers = 0;
    for(let key in reviewObject) {
      sumTotal += key * reviewObject[key];
      sumReviewers += Number(reviewObject[key]);
    }
    return sumTotal/sumReviewers;
  };
  const rating = averageRating();
  // const [image, setImage] = useState('./testImage.png');
  return (
    <CardStyled onClick={handleCardClick}>
      <ComparisonModal modalIsOpen={modalIsOpen} currentProduct={curProduct} comparisonProduct={product.data}/>
      <ProductImageStyled >
        <ImageStyled src={defaultImage} alt="Image of RelatedProduct" aria-label="Product Image"/>
        <ButtonStyled src={yellowStar}
          onMouseEnter={handleOnMouseEnter}
          onMouseLeave={handleOnMouseLeave}
          aria-label="Comparison Button"/>
        <SmallStarBar rating={rating}/>
      </ProductImageStyled>
      <OuterDescriptionDiv>
        <Category>{product.data.category}</Category>
        <ProductTitle>{product.data.name}</ProductTitle>
        <Price>${product.data.default_price}</Price>
      </OuterDescriptionDiv>
    </CardStyled>
  )
}

export default ProductCard;