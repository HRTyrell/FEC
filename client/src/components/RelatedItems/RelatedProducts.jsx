import {useState, useEffect} from 'react';
import ProductCard from './ProductCard.jsx';
import styled from 'styled-components';

const Div = styled.div`
display: flex;
flex-direction: row;
border-spacing: 20px;
padding: 0;
`;
const Title = styled.h5`

`;

const RelatedProducts = ({data}) => {

  return (
    <Div>
      {data.map((product, index) => {
        return <ProductCard product={product} key={product.data.id}/>
      })}
    </Div>
  )
};

export default RelatedProducts;