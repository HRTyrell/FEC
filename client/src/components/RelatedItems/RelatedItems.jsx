import React, {useState, useEffect, useRef} from 'react';
import getProducts from './parseHelpers.js';
import RelatedProducts from './RelatedProducts.jsx';
import YourOutfit from './YourOutfit.jsx';
import styled from 'styled-components';
import Carousel from './Carousel.jsx';
import testData from './TestData.js';
import {getRelatedProducts} from './parseHelpers.js';
import {useRelatedItemsStore} from './RelatedItemsStore.jsx';
import ProductStore from '../Provider/Zus_Provider.jsx';

const Title = styled.h1`

`;
const Div = styled.div`
  display: flex;
  flex-direction: column;
  align-content: space-between;
`;
const MasterDiv = styled.div`
  display: flex;
  flex-direction: row;
  align-content: space-between;
  width: 80%;
  border-spacing: 0.5%;
  margin: 0;
  padding: 0;
  line-height: 0.8;
`;
const H = styled.h5`
  display: flex;
  align-items: center;
  justify-content: center;
  height: 2em;
  margin-block-start: 2em;
  margin-block-end: 0;
`;

const RelatedItems = () => {

  const setRelatedItems = useRelatedItemsStore(state => state.setRelatedItems);
  const {curProduct} = ProductStore();
  const isMounted = useRef(false);

  useEffect(() => {
    if(isMounted.current) {
      getRelatedProducts(curProduct.id).then(items => {
        setRelatedItems(items);
      })
    } else {
      isMounted.current = true;
    }
  }, [curProduct])
  const { relatedItemsList, outfitList} = useRelatedItemsStore();

  return (
    <Div>
      <H>Related Products</H>
      <MasterDiv>
        <Carousel data={relatedItemsList} title="Related Products"><RelatedProducts/></Carousel>
      </MasterDiv>
      <H>Your Outfit</H>
      <MasterDiv>
        <Carousel data={outfitList} title="Your Outfit"><YourOutfit/></Carousel>
      </MasterDiv>
    </Div>

  )
}

export default RelatedItems;