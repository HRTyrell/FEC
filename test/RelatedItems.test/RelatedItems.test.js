import RelatedItems from '../../client/src/components/RelatedItems/RelatedItems.jsx';
import renderer from 'react-test-renderer';

test('Test Renderer', () => {

  const testRender = renderer.create(<RelatedItems />)
  expect(parseHelpers).not.toBeNull();
})