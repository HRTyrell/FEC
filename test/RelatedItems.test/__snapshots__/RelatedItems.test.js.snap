// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`next hides on change 1`] = `
<div
  style={
    {
      "maxWidth": "500px",
      "position": "relative",
    }
  }
>
  <div
    className="sc-eDvSVe htMdsl"
  />
  <div
    aria-label="Product Carousel"
    className="sc-bcXHqe dIXrQb"
  >
    <button
      className="sc-gswNZR sc-hLBbgP gDWjNX kbQPoy"
      onClick={[Function]}
    >
      &gt;
    </button>
    <div
      className="sc-eDWCr FgxmC"
    >
      <div
        className="sc-jSUZER jqkdYo"
      >
        <div
          className="sc-gKPRtg kHPEnk"
        >
          <img
            alt="Image of RelatedProduct"
            className="sc-iBYQkv iHGUpD"
            src="./testImage.png"
          />
          <img
            aria-label="Comparison Button"
            className="sc-ftTHYK bwbWm"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            src="./star-empty-icon.png"
          />
        </div>
        <div
          className="sc-pyfCe gajNgd"
        >
          <p
            className="sc-jrcTuL lcbUeq"
          >
            Jackets
          </p>
          <h5
            className="sc-kDvujY bkUuDT"
          >
            Camo Onesie
          </h5>
          <p
            className="sc-ipEyDJ jAscTz"
          >
            140
          </p>
          <img
            className="sc-csuSiG hlkPKg"
            src=""
          />
        </div>
      </div>
      <div
        className="sc-jSUZER jqkdYo"
      >
        <div
          className="sc-gKPRtg kHPEnk"
        >
          <img
            alt="Image of RelatedProduct"
            className="sc-iBYQkv iHGUpD"
            src="./testImage.png"
          />
          <img
            aria-label="Comparison Button"
            className="sc-ftTHYK bwbWm"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            src="./star-empty-icon.png"
          />
        </div>
        <div
          className="sc-pyfCe gajNgd"
        >
          <p
            className="sc-jrcTuL lcbUeq"
          >
            Accessories
          </p>
          <h5
            className="sc-kDvujY bkUuDT"
          >
            Bright Future Sunglasses
          </h5>
          <p
            className="sc-ipEyDJ jAscTz"
          >
            69
          </p>
          <img
            className="sc-csuSiG hlkPKg"
            src=""
          />
        </div>
      </div>
      <div
        className="sc-jSUZER jqkdYo"
      >
        <div
          className="sc-gKPRtg kHPEnk"
        >
          <img
            alt="Image of RelatedProduct"
            className="sc-iBYQkv iHGUpD"
            src="./testImage.png"
          />
          <img
            aria-label="Comparison Button"
            className="sc-ftTHYK bwbWm"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            src="./star-empty-icon.png"
          />
        </div>
        <div
          className="sc-pyfCe gajNgd"
        >
          <p
            className="sc-jrcTuL lcbUeq"
          >
            Pants
          </p>
          <h5
            className="sc-kDvujY bkUuDT"
          >
            Morning Joggers
          </h5>
          <p
            className="sc-ipEyDJ jAscTz"
          >
            40
          </p>
          <img
            className="sc-csuSiG hlkPKg"
            src=""
          />
        </div>
      </div>
      <div
        className="sc-jSUZER jqkdYo"
      >
        <div
          className="sc-gKPRtg kHPEnk"
        >
          <img
            alt="Image of RelatedProduct"
            className="sc-iBYQkv iHGUpD"
            src="./testImage.png"
          />
          <img
            aria-label="Comparison Button"
            className="sc-ftTHYK bwbWm"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            src="./star-empty-icon.png"
          />
        </div>
        <div
          className="sc-pyfCe gajNgd"
        >
          <p
            className="sc-jrcTuL lcbUeq"
          >
            Jackets
          </p>
          <h5
            className="sc-kDvujY bkUuDT"
          >
            Camo Onesie
          </h5>
          <p
            className="sc-ipEyDJ jAscTz"
          >
            140
          </p>
          <img
            className="sc-csuSiG hlkPKg"
            src=""
          />
        </div>
      </div>
      <div
        className="sc-jSUZER jqkdYo"
      >
        <div
          className="sc-gKPRtg kHPEnk"
        >
          <img
            alt="Image of RelatedProduct"
            className="sc-iBYQkv iHGUpD"
            src="./testImage.png"
          />
          <img
            aria-label="Comparison Button"
            className="sc-ftTHYK bwbWm"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            src="./star-empty-icon.png"
          />
        </div>
        <div
          className="sc-pyfCe gajNgd"
        >
          <p
            className="sc-jrcTuL lcbUeq"
          >
            Accessories
          </p>
          <h5
            className="sc-kDvujY bkUuDT"
          >
            Bright Future Sunglasses
          </h5>
          <p
            className="sc-ipEyDJ jAscTz"
          >
            69
          </p>
          <img
            className="sc-csuSiG hlkPKg"
            src=""
          />
        </div>
      </div>
      <div
        className="sc-jSUZER jqkdYo"
      >
        <div
          className="sc-gKPRtg kHPEnk"
        >
          <img
            alt="Image of RelatedProduct"
            className="sc-iBYQkv iHGUpD"
            src="./testImage.png"
          />
          <img
            aria-label="Comparison Button"
            className="sc-ftTHYK bwbWm"
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
            src="./star-empty-icon.png"
          />
        </div>
        <div
          className="sc-pyfCe gajNgd"
        >
          <p
            className="sc-jrcTuL lcbUeq"
          >
            Pants
          </p>
          <h5
            className="sc-kDvujY bkUuDT"
          >
            Morning Joggers
          </h5>
          <p
            className="sc-ipEyDJ jAscTz"
          >
            40
          </p>
          <img
            className="sc-csuSiG hlkPKg"
            src=""
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
